[HOLD] Blazor WASM debug topic updates for 3.2p3 (#17438)
dotnet/AspNetCore.Docs: aspnetcore/blazor/debug.md

ms.author: riande
ms.custom: mvc
ms.date: 03/24/2020
no-loc: [Blazor, SignalR]
uid: blazor/debug
---
# Debug ASP.NET Core Blazor WebAssembly

[Daniel Roth](https://github.com/danroth27)

[!INCLUDE[](~/includes/blazorwasm-preview-notice.md)]

Blazor WebAssembly apps can be debugged using the browser dev tools in Chromium-based browsers (Edge/Chrome).  Alternatively you can debug your app using Visual Studio or Visual Studio Code.

Available scenarios include:

* Set and remove breakpoints.
* Run the app with debugging support in Visual Studio and Visual Studio Code (<kbd>F5</kbd> support).
* Single-step (<kbd>F10</kbd>) through the code.
* Resume code execution with <kbd>F8</kbd> in a browser or <kbd>F5</kbd> in Visual Studio or Visual Studio Code.
* In the *Locals* display, observe the values of local variables.
* See the call stack, including call chains that go from JavaScript into .NET and from .NET to JavaScript.

You *can't*:

* Inspect arrays.
* Hover to inspect members.
* Step debug into or out of managed code.
* Have full support for inspecting value types.
* Break on unhandled exceptions.
* Hit breakpoints during app startup.
* Debug an app with a service worker.

> We expect to continue to improve the debugging experience in future releases.

Reviewer: My own perception of "future" is a bit vague and it leaves me with an impression that this may not happen.
What about slightly modified version below (please ignore this, if you find your phrasing better):

```suggestion
We will continue to improve the debugging experience in upcoming releases.
```

## Prerequisites

Debugging requires either of the following browsers:

* Microsoft Edge (version 80 or later)
* Google Chrome (version 70 or later)

> ## Enable debugging for Visual Studio and Visual Studio Code

Reviewer: Should this rather be 
```suggestion
## Enable debugging for existing Blazor WebAssembly projects

Debugging should work out of the box for projects created using 3.2-preview3 and later project templates.
For existing projects, , however, a gesture is required.
```

Author: No, it should say that explicitly because the following instruction doesn't apply to debugging in the browser. The real problem here is that we can't have **_custom tabs_**. If we could have custom tabs, then this content could go **only** at the tops of the VS and VSC tabs. The 'Debug in the browser' tab wouldn't get it.

Reviewer: Ah, good point! I think the title can remain in that case. You're choice whether the other sentence is worthy or not

Author: Yes ... I'll fix that now. My :eye: went straight to the highlighted text in the diff, so I didn't pick up on the rest. I'll fix it now.

---
title: Debug ASP.NET Core Blazor WebAssembly
author: guardrex
description: Learn how to debug Blazor apps.
monikerRange: '>= aspnetcore-3.1'
ms.author: riande
ms.custom: mvc
ms.date: 03/24/2020
no-loc: [Blazor, SignalR]
uid: blazor/debug
---
# Debug ASP.NET Core Blazor WebAssembly

[Daniel Roth](https://github.com/danroth27)

[!INCLUDE[](~/includes/blazorwasm-preview-notice.md)]

Blazor WebAssembly apps can be debugged using the browser dev tools in Chromium-based browsers (Edge/Chrome).  Alternatively you can debug your app using Visual Studio or Visual Studio Code.

Available scenarios include:

* Set and remove breakpoints.
* Run the app with debugging support in Visual Studio and Visual Studio Code (<kbd>F5</kbd> support).
* Single-step (<kbd>F10</kbd>) through the code.
* Resume code execution in a browser with <kbd>F8</kbd> or <kbd>F5</kbd> in Visual Studio or Visual Studio Code.

> * In the *Locals* display, observe the values of any local variables.

Reviewer 2: ```suggestion
* In the *Locals* display, observe the values of local variables.
```

I think we still have some issues viewing certain value types.

Once updated, the *launchSettings.json* file should look something like the following:

```json
{
  "iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:50454",
      "sslPort": 44399
    }
  },
  "profiles": {
    "IIS Express": {
      "commandName": "IISExpress",
      "launchBrowser": true,
      "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    },
    "BlazorApp1.Server": {
      "commandName": "Project",
      "launchBrowser": true,
      "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
      "applicationUrl": "https://localhost:5001;http://localhost:5000",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}
```

The `inspectUri` property:

* Enables the IDE to detect that the app is a Blazor WebAssembly app.
* Instructs the script debugging infrastructure to connect to the browser through Blazor's debugging proxy.

> ## Procedure

Reviewer 2: Instead of having "Procedure" as section that shows up in the right nav, can we remove the Procedure heading and elevate the following sections to be h2s?

     ![Additional setup requried](https://devblogs.microsoft.com/aspnet/wp-content/uploads/sites/16/2020/03/vscode-additional-setup.png)

   * A notification offers to add the required assets to the app for building and debugging. Select **Yes**:

     ![Add required assets](https://devblogs.microsoft.com/aspnet/wp-content/uploads/sites/16/2020/03/vscode-required-assets.png)

1. Starting the app in the debugger is a two-step process:

   1\. **First**, start the app using the **.NET Core Launch (Blazor Standalone)** launch configuration.

   2\. **After the app has started**, start the browser using the **.NET Core Debug Blazor Web Assembly in Chrome** launch configuration (requires Chrome). To use Edge instead of Chrome, change the `type` of the launch configuration in *.vscode/launch.json* from `pwa-chrome` to `pwa-edge`.

1. Set a breakpoint in the `IncrementCount` method in the `Counter` component and then select the button to hit the breakpoint:

   ![Debug Counter in VS Code](https://devblogs.microsoft.com/aspnet/wp-content/uploads/sites/16/2020/03/vscode-debug-counter.png)

> ### Debug in the browser

Reviewer 2: It loos like the "Enable remote debugging" and "Debug the app" sections below should just be merged into this section as both are about in-browser debugging.

1. Run a Debug build of the app in the Development environment.

1. Press <kbd>Shift</kbd>+<kbd>Alt</kbd>+<kbd>D</kbd>.

1. Follow the instructions to restart the browser with [remote debugging enabled](#enable-remote-debugging), if needed.

1. Use the browser's developer tools to debug the app.

## Enable remote debugging

> If remote debugging is disabled, an **Unable to find debuggable browser tab** error page is generated by Chrome. The error page contains instructions for running Chrome with the debugging port open so that the Blazor debugging proxy can connect to the app. *Close all Chrome instances* and restart Chrome as instructed.

Reviewer 2: ```suggestion
To debug a Blazor WebAssembly app the browser must be run with remote debugging enabled. If remote debugging is disabled, an **Unable to find debuggable browser tab** error page is generated. The error page contains instructions for running the browser with the debugging port open so that the Blazor debugging proxy can connect to the app. *Close all browser instances* and restart the browser as instructed.
```

[!INCLUDE[](~/includes/blazorwasm-preview-notice.md)]

> As of ASP.NET Core 3.2 Preview 3, Blazor WebAssembly apps can be debugged using the browser dev tools in Chromium-based browsers (Chrome/Edge) in Visual Studio and Visual Studio Code.

Reviewer 2: ```suggestion
Blazor WebAssembly apps can be debugged using the browser dev tools in Chromium-based browsers (Edge/Chrome).  Alternatively you can debug your app using Visual Studio or Visual Studio Code.
```

Available scenarios include:

* Set and remove breakpoints.
* Run the app with debugging support in Visual Studio (<kbd>F5</kbd> support).

> * Single-step (<kbd>F10</kbd>) through the code or resume (<kbd>F8</kbd>) code execution.

Reviewer 2: F8 is only used for resuming execution in a browser. In VS or VS Code you'd use F5.

> * Google Chrome (version 70 or later)

Reviewer 2: Let's switch the order and put Edge first.

> * Microsoft Edge preview ([Edge Dev Channel](https://www.microsoftedgeinsider.com))

Reviewer 2: The Chromium based Microsoft Edge is no longer in preview:  https://microsoft.com/edge. If you have Edge version 80 or later, you should be good to go.

> ## Enable debugging

Reviewer 2: Adding `inspectUri` is only needed for debugging in our IDEs. If you just want to debug in the browser, then this isn't really required.

In addition to debugging in our IDEs, we still need cover how to debug in a browser using the browser dev tools.
1. Run a Debug build of the app in the Development environment
1. Hit <kbd>Shift-Alt-D</kbd>
1. Follow the instructions to restart the browser with remote debugging enabled if needed.
1. Use the browser dev tools to debug your app

Then we can cover debugging in an IDE

> To enable debugging in an existing Blazor WebAssembly app, update *launchSettings.json* in the startup project of your app. Set `useWebAssemlbyDebugging` to `true` and include the following `inspectUri` property in each launch profile:

Reviewer 2: Turns out `useWebAssemblyDebugging` isn't necessary and shouldn't be in the template at all.

```suggestion
To enable debugging in an existing Blazor WebAssembly app in Visual Studio or Visual Studio Code, update *launchSettings.json* in the startup project of your app to include the following `inspectUri` property in each launch profile:
```

 
```json
"inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}"
```
 
Once updated, the *launchSettings.json* file should look something like the following:
 
```json
{

>   "useWebAssemblyDebugging": true,

Reviewer 2: The `useWebAssemblyDebugging` property actually isn't needed. It still shows up in our templates, but we'll get that cleaned up in the next release.

```suggestion
```

  "iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:50454",
      "sslPort": 44399
    }
  },
  "profiles": {
    "IIS Express": {
      "commandName": "IISExpress",
      "launchBrowser": true,
      "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    },
    "BlazorApp1.Server": {
      "commandName": "Project",
      "launchBrowser": true,
      "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
      "applicationUrl": "https://localhost:5001;http://localhost:5000",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    }
  }
}
```

The preceding properties:

* Enable the IDE to detect that the app is a Blazor WebAssembly app.

> * Instruct the script debugging infrastructure to connect to the browser through Blazor's debugging proxy.

Reviewer 2: ```suggestion
The `inspectUri property:

* Enables the IDE to detect that the app is a Blazor WebAssembly app.
* Instructs the script debugging infrastructure to connect to the browser through Blazor's debugging proxy.
```

## Debug the app

> Once Chrome is running with remote debugging enabled, the debugging keyboard shortcut opens a new debugger tab. After a moment, the **Sources** tab shows a list of the .NET assemblies in the app. Expand each assembly and find the *.cs*/*.razor* source files available for debugging. Set breakpoints, switch back to the app's tab, and the breakpoints are hit when the code executes. After a breakpoint is hit, single-step (<kbd>F10</kbd>) through the code or resume (<kbd>F8</kbd>) code execution normally.

Reviewer 2: ```suggestion
Once the browser is running with remote debugging enabled, the debugging keyboard shortcut opens a new debugger tab. After a moment, the **Sources** tab shows a list of the .NET assemblies in the app. Expand each assembly and find the *.cs*/*.razor* source files available for debugging. Set breakpoints, switch back to the app's tab, and the breakpoints are hit when the code executes. After a breakpoint is hit, single-step (<kbd>F10</kbd>) through the code or resume (<kbd>F8</kbd>) code execution normally.
```
